### ISSUE TITLE
Emit typescript definition file for #[js_export] functions

### ISSUE THREAD

**Issue by bgourlie** [koute/cargo-web#65]
I found myself manually creating typescript definition files for my exported rust functions because my IDE will use them for static analysis. Then I thought it would be neat if typescript definition were emitted automatically. 

I can take a stab at implementing this. I'm pretty unfamiliar with the code base. Where would this functionality live?

**Comment by Pauan** [comment]
@bgourlie I'm curious: what do your TypeScript definition files look like? I would expect them to all be functions that accept `number` and return `number`

**Comment by bgourlie** [comment]
So for my first web assembly project I'm making an interactive float tutorial (enter a float, see the discrete parts that make up the value, show how they make up the final value, precision, etc...).

For the front end, I'm using React written in TypeScript (tsx) with `noImplicitAny` set to true, giving me strict type checking. The interfaces declared in the `ts.d` file map directly to the json serializable types defined in my WASM module, likewise with any #[js_export] functions declarations.

```
declare module 'interactive-float-tutorial' {
    export interface FloatValue {
        sign_bit: boolean,
        exponent_bits: [boolean],
        exponent_value: number,
        mantissa_bits: [boolean],
        mantissa_value: number,
        original_value: number
    }

    export function default_float_value(): FloatValue;
    export function update(val: Float64): void;
}

```

**Comment by Pauan** [comment]
@bgourlie I see, thanks!

**Comment by koute** [comment]
@Pauan As @bgourlie explained it's not about the *raw* functions exported from the `.wasm` file but about the properly-typed wrappers around those functions generated with `#[js_export]`.

@bgourlie That would require modifyng both `#[js_export]` (which lives inside the `stdweb-internal-macros` crate) and `cargo-web` (more specifically `wasm_js_export.rs`). The `TypeMetadata::Custom` structure has a `name` field which is currently unused, but which I've meant to eventually use for something like this.

**Comment by Pauan** [comment]
@koute Yup, I understand that now. It does sound useful to generate TypeScript `.d.ts` for the wrappers.
